File: Leetcode/Python/116. Populating Next Right Pointers in Each Node/solution.py
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node/

# Ref: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/discuss/1654181/C%2B%2BPythonJava-Simple-Solution-w-Images-and-Explanation-or-BFS-%2B-DFS-%2B-O(1)-Optimized-BFS

"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

from collections import deque


class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root == None:
            return None
        q = deque([root])
        # Fill next node from the right to the left to put null in the rightmost node
        while q:
            nextNode = None
            for _ in range(len(q)):
                now = q.popleft()
                now.next, nextNode = nextNode, now
                if now.right:
                    q.extend([now.right, now.left])
        return root

File: Leetcode/Typescript/116. Populating Next Right Pointers in Each Node/solution.ts
// https://leetcode.com/problems/populating-next-right-pointers-in-each-node/description/

/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     left: _Node | null
 *     right: _Node | null
 *     next: _Node | null
 *     constructor(val?: number, left?: _Node, right?: _Node, next?: _Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function connect(root: _Node | null): _Node | null {
  if (!root) {
    return null;
  }

  const linkNode: _Node[] = [];

  function search(root: _Node | null, level: number) {
    if (!root) {
      return;
    }

    if (linkNode.length === level) {
      linkNode.push(new _Node());
    }
    linkNode[level].next = root;
    linkNode[level] = root;

    search(root.left, level + 1);
    search(root.right, level + 1);
  }

  search(root, 0);

  return root;
}
